File: src/y04.js
// Higher Order Functions
// Functional Refactoring
// (1) Tennentt Correspondence Principle
// (2) Introduce new binding
// (3) Rebind existing binding
// (4) Inline function definition

error = function(n) { throw Error("SHOULD NOT BE CALLED") }

fact_improver = function(partial) {
  return function(n) {
    return n==0 ? 1 : n * partial(n-1)
  }
}

y = function(g) {
  return function(f) { return function(n) { return g(f(f))(n) } }(
         function(f) { return function(n) { return g(f(f))(n) } })
}

fact = y(fact_improver)

result = fact(5)

console.log(result);

File: src/y04.clj
(println
 (do

   ;; Higher Order Functions
   ;; Refactorings
   ;; (1) Inlining
   ;; (2) Function Wrapping
   ;; Fixpoint

   (def error (fn [n] (throw (Throwable. "SHOULD NOT CALL"))))

   (def fact-partial
        (fn [recursive]
          (fn [n] (if (zero? n) 1 (* n (recursive (- n 1)))))))

   ;; Applicative Order Y-Combinator

   (def y
        (fn [f]
          ((fn [fp] (fn [n] ((f (fp fp)) n)))
           (fn [fp] (fn [n] ((f (fp fp)) n))))))

   (def fact (y fact-partial))

   (fact 5)

))
